wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileIntegrityValidator.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.devfile.server;
+
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toSet;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.regex.Pattern;
+import javax.inject.Singleton;
+import org.eclipse.che.api.devfile.model.Action;
+import org.eclipse.che.api.devfile.model.Command;
+import org.eclipse.che.api.devfile.model.Devfile;
+import org.eclipse.che.api.devfile.model.Project;
+import org.eclipse.che.api.devfile.model.Tool;
+
+/**
+ * Validates devfile logical integrity. Performs the following checks:
+ *
+ * <pre>
+ * <ul>
+ *   <li>All listed items (projects, tools, commands) have unique names</li>
+ *   <li>There is only one tool of type cheEditor</li>
+ *   <li>All tools exists which are referenced/required by command actions</li>
+ *   <li>Project names conforms naming rules</li>
+ * </ul>
+ * </pre>
+ */
+@Singleton
+public class DevfileIntegrityValidator {
+
+  private static final Pattern PROJECT_NAME_PATTERN = Pattern.compile("^[\\w\\d]+[\\w\\d_.-]*$");
+
+  void validateDevfile(Devfile devfile) throws DevfileFormatException {
+    validateProjects(devfile);
+    Set<String> toolNames = validateTools(devfile);
+    validateCommands(devfile, toolNames);
+  }
+
+  private Set<String> validateTools(Devfile devfile) throws DevfileFormatException {
+    Set<String> existingNames = new HashSet<>();
+    Tool editorTool = null;
+    for (Tool tool : devfile.getTools()) {
+      if (!existingNames.add(tool.getName())) {
+        throw new DevfileFormatException(format("Duplicate tool name found:'%s'", tool.getName()));
+      }
+      switch (tool.getType()) {
+        case "cheEditor":
```

Comment:
Is there a corresponding constant for `cheEditor` tool type?

Reply:
Added